Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -15,709 +15,573 @@
-            return buffer
+    private func didReceiveHeaderFieldData(_ bytes: UnsafeRawBufferPointer) {
+        switch self.decodingState {
+        case .headerName, .trailerName:
+            ()
+        case .headerValue:
+            self.finish { decoder, bytes in
+                decoder.didReceiveHeaderValue(bytes)
+            }
+            self.start(bytes: bytes, newState: .headerName)
+        case .trailerValue:
+            self.finish { decoder, bytes in
+                decoder.didReceiveTrailerValue(bytes)
+            }
+            self.start(bytes: bytes, newState: .trailerName)
+        case .url:
+            self.finish { decoder, bytes in
+                decoder.didReceiveURL(bytes)
+            }
+            self.start(bytes: bytes, newState: .headerName)
+        case .headersComplete:
+            // these are trailers
+            self.start(bytes: bytes, newState: .trailerName)
+        case .beforeURL:
+            // in case we're parsing responses
+            self.start(bytes: bytes, newState: .headerName)
+        default:
+            preconditionFailure()
         }
-        return nil
+        self.currentFieldByteLength += bytes.count
     }
 
-    public func handlerRemoved(context: ChannelHandlerContext) {
-        if let buffer = self.bytesToForwardOnRemoval(context: context) {
-            context.fireChannelRead(NIOAny(buffer))
+    private func didReceiveHeaderValueData(_ bytes: UnsafeRawBufferPointer) {
+        switch self.decodingState {
+        case .headerValue, .trailerValue:
+            ()
+        case .headerName:
+            self.finish { decoder, bytes in
+                decoder.didReceiveHeaderName(bytes)
+            }
+            self.start(bytes: bytes, newState: .headerValue)
+        case .trailerName:
+            self.finish { decoder, bytes in
+                decoder.didReceiveTrailerName(bytes)
+            }
+            self.start(bytes: bytes, newState: .trailerValue)
+        default:
+            preconditionFailure()
         }
-        self.cumulationBuffer = nil
+        self.currentFieldByteLength += bytes.count
     }
-    
-    public func handlerAdded(context: ChannelHandlerContext) {
-        if HTTPMessageT.self == HTTPServerRequestPart.self {
-            c_nio_http_parser_init(&self.parser, HTTP_REQUEST)
-        } else if HTTPMessageT.self == HTTPClientResponsePart.self {
-            c_nio_http_parser_init(&self.parser, HTTP_RESPONSE)
-        } else {
-            fatalError("the impossible happened: MsgT neither HTTPClientRequestPart nor HTTPClientResponsePart but \(HTTPMessageT.self)")
-        }
-
-        self.parser.data = Unmanaged.passUnretained(context).toOpaque()
-
-        c_nio_http_parser_settings_init(&self.settings)
 
-        self.settings.on_message_begin = { parser in
-            let handler = evacuateHTTPDecoder(parser)
-            handler.state.reset()
+    private func didReceiveStatusData(_ bytes: UnsafeRawBufferPointer) {
+    }
```

Comment:
I think it'd be nice to put a comment here saying why you don't need a method body here.